### DISCUSSION TITLE
ESP32 SDCard OSError: 16

### DISCUSSION THREAD

**Discussion by engineertrooper** [micropython/micropython#10581] [ESP32] [answered]
Hi Community,

**Problem:**
I get an OSError: 16 if I use the example code [V1.19.1](https://docs.micropython.org/en/v1.19.1/library/machine.SDCard.html#machine-sdcard) for the SDCard Class.

**My Device:**
- ESP32 Dev Kit C V2 or also sometimes named ESP32 NodeMCU Module
- with Micropython: v1.19.1 -> 2022-06-18
- Pins: sck=18, miso=19, mosi=23, cs=5 and 3,3V supply

**Pre-Analyse**
An OSError: 16 means a device or ressource is busy. Because of that I commented out all code with exception the following lines:

```python

import os
import machine

sd = machine.SDCard(slot=2)
os.mount(sd, "/sd")  # mount

os.listdir('/sd')    # list directory contents

```

Error-Msg (always the same, it doesn't matter if other code is commented out or not):

```

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
Traceback (most recent call last):
  File "main.py", line 62, in <module>
OSError: 16
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>> 

```

**Looking into the world wide web, github issues and other discussions**

While I was reading I got the feeling that it has not something to do with the board directly. It looks like to me that the error is inside the ESP-IDF or there was a change in Micropython that was making that error existing?

After reading some Issues, Documentation, Discussions and Tutorials I marked down the following links:

- [Class SDCard](https://docs.micropython.org/en/v1.19.1/library/machine.SDCard.html#machine-sdcard)
- [Class SPI](https://docs.micropython.org/en/v1.9.1/pyboard/library/machine.SPI.html)
- [MicroPython Forum](https://forum.micropython.org/viewtopic.php?t=12007)
- [#8949](https://github.com/micropython/micropython/issues/8949)
- [#8290](https://github.com/micropython/micropython/issues/8290)
- [#9925](https://github.com/micropython/micropython/discussions/9935)
- [#429 micropython-lib](https://github.com/micropython/micropython-lib/issues/429)
- [#8940](https://github.com/micropython/micropython/issues/8940)

**Note: tutorials excluded, because old solutions with a [blob](https://github.com/micropython/micropython/blob/master/drivers/sdcard/sdcard.py) from @dpgeorge was used. But I found something similar of the blob [here](https://github.com/micropython/micropython-infineon/blob/master/drivers/sdcard/sdcard.py).*

***Issue 429 and 8940 maybe not related to this question, but I'm not very fimilar to ESP-IDF, so added them just in case.*

**SD-Card Frequenz**

One try was to set a frequenz. -> ``` freq=20000000 ``` as parameter inside ```SDCard(...)```
I played with diffrent frequenzes between 100KHz and 20MHz, but didn't work. (See [#8290 comment](https://github.com/micropython/micropython/issues/8290#issuecomment-1054208392))

*Note: I'm using a 32GB SanDisk Ultra MicroSdCard.*

**Question:**

Any Idea how this **OSError: 16** is coming up?
Or Do I oversee something simple here?

With these Information I wanted to ask here first, before opening a new issue.

**Comment by BNNorman** [comment]
try this:-
```
sd=machine.SDCard(slot=2)
vfs=os.VfsFat(sd)
os.mount(sd,"/sd")
```

However, I have had problems with slot=2 (and slot=3). If the code is re-run ,without pressing Stop in Thonny, I get
```
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
OSError: (-259, 'ESP_ERR_INVALID_STATE')
```

This can be repeated with a very very simple program:-

```
import machine

sd=machine.SDCard(slot=2)
```
So, as long as your code runs immediately after a reboot it should run fine but it's a pig during development with REPL.

It looks like a bug in "MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32" to me. The SDCard isn't deleted properly when the program ends. I'm switching to a Pico & CircuitPython  for my logger code, maybe I'll have more luck there.

**Reply by engineertrooper** [reply]
Hi, thanks for the reply.

I tested your suggestion. But I get the error as well.
Here is the Code and the Errors-Msg:

```python

import os
import machine

sd = machine.SDCard(slot=2)
vfs=os.VfsFat(sd)
os.mount(sd, "/sd")  # mount

os.listdir('/sd')    # list directory contents

```

the error-msg:

```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
Traceback (most recent call last):
  File "main.py", line 62, in <module>
OSError: 16
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>> 
```

**OSError: [Errno 2] ENOENT**

I want to point out a other try here as well. In the [documentation to vfs](https://docs.micropython.org/en/v1.19/reference/filesystem.html#vfs) is written "On ESP8266/ESP32, the primary filesystem is mounted at ```'/'```." So I assumed the the SD-Card can be accessed with ```/sd```. This is not working as well. The SD-Card is not mounted even by default. That default is only for STM32/Pyboard.

I tryed some other stuff more, according to the documentation at v1.19 #vfs. But didn't worked out.

**Solution as workaround with a module "sdcard.py" in 2021 ?**

Even this is not working. But it was my last idea anyway. I found the following links to that.

- [Micropython Forum](https://forum.micropython.org/viewtopic.php?t=11205)
- [#7352](https://github.com/micropython/micropython/issues/7352)
- [#7414](https://github.com/micropython/micropython/issues/7414)
- [az-delivery.de -> Tutorial (in German)](https://www.az-delivery.de/en/blogs/azdelivery-blog-fur-arduino-und-raspberry-pi/micropython-und-esp32-esp8266-als-datenlogger)
  - I tryed this [grzesina.de -> sdcard.py](http://www.grzesina.de/az/datalogger/sdcard.py) file.
  
  
Testcode: Trying to unmount and mount the sd card isn't working. I also tested to not unmount the system, same result.

```python
import os
import machine
import sdcard           #external sdcard.py load from '/'

print(os.listdir('/'))

os.umount('/')
spi = machine.SPI(2,baudrate=100000,sck=machine.Pin(18),mosi=machine.Pin(23),
       miso=machine.Pin(19),polarity=0,phase=0)
sd = sdcard.SDCard(spi, machine.Pin(5))
try:
   os.mount(sd, '/sd')
   print("SD-Card is mounted on /sd")
except OSError as e:
   print(e)
   print("SD-Card previously mounted")
```

I get the following error-msg:

```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
['boot.py', 'main.py', 'sdcard.py']
Traceback (most recent call last):
  File "main.py", line 76, in <module>
  File "sdcard.py", line 47, in __init__
  File "sdcard.py", line 72, in init_card
OSError: no SD card
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>> 
```

**Please note:**

*I don't want to create a "fix" by using an external sdcard.py file. Then it wouldn't make sense to open an issue to this discussion. And also micropython should grow* 😄

-------------

I will open up a new issue now. Let's see what the community says to this. I think it is a bug in micropython itself on version v1.19.1 on 2022-06-18 for the ESP32 Dev Kit C V2.

**Reply by engineertrooper** [reply]
Issue opened: [https://github.com/micropython/micropython/issues/10634](https://github.com/micropython/micropython/issues/10634)

**Comment by robert-hh** [comment]
Try:
```
import os
import machine

sd = machine.SDCard(slot=2)
vfs=os.VfsFat(sd)
os.mount(vfs, "/sd")  # mount

os.listdir('/sd')    # list directory contents
```
The difference is in the mount command, where it uses the vfs as first argument.

Besides that, sdcard.py is a little bit picky about SD card brand, layout & sizes. The card has to have a single FAT partition only, and I used it with cards in the size range from 8GB to 64GB, brands Samsung, Sandisk and Transcend.

**Reply by engineertrooper** [reply]
Hi thank you for idea.
tryed the code.

Error-Msg stays the same:
```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12344
ho 0 tail 12 room 4
load:0x40080400,len:4124
entry 0x40080680
Traceback (most recent call last):
  File "main.py", line 5, in <module>
OSError: 16
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
```

It looks like the problem is connected to ```vfs=os.VfsFat(sd)```.

-----------
*Besides that, sdcard.py is a little bit picky about SD card brand, layout & sizes. The card has to have a single FAT partition only, and I used it with cards in the size range from 8GB to 64GB, brands Samsung, Sandisk and Transcend.*

-> Yes, but lucky I have a 32GB SanDisk Ultra MicroSdCard. I think that should work.

**Reply by robert-hh** [reply]
The error code is generated by the VfsFat driver if the device is not ready. So It still looks as if there is no access to the card or the card is not detected.
The card brand & size is fine. Did you check that it has only a single partition, just for the sake for simplicity?

**Reply by BNNorman** [reply]
Odd because machine.SDCard works on my nodemcu esp32s. I used an 8gb samsung and a 32gb sandisk Both formatted fat32 and mounted on /sd. My sd device is just a cheap thing off ebay.

The sd card code worked for me it just failed whenever I tried to rerun it with REPL without pressing Stop in Thonny first.

I wonder if the upload of the UF2 was corrupted. Might be worth erasing flash then reflash the UF2?

**Reply by engineertrooper** [reply]
> The error code is generated by the VfsFat driver if the device is not ready. So It still looks as if there is no access to the card or the card is not detected. The card brand & size is fine. Did you check that it has only a single partition, just for the sake for simplicity?

Yes, I checked and format FAT32 as well again. Error still there.
On Win11 I used the system build in Format with
- File system: FAT32 (Default)
- allocation unit size: 32 kilobytes
- Volume label: <no name> (the field is empty!)
- Quick Format unchecked

**Reply by engineertrooper** [reply]
> reflash the UF2

I haven't done this before. Would be my first time. How I do that with my ESP32 Dev Kit C V2 ?
The Internet shows some tutorials but do you have a specific way in mind, @BNNorman ?

**Comment by BNNorman** [comment]
How did you put micropython on your esp32 before?

Basically, use esptool. For doing it on Windows there's this tutorial

 https://microcontrollerslab.com/flash-micropython-firmware-esptool-py-esp32-esp8266/#:~:text=The%20esptool.py%20tools%20can%20be%20used%20on%20Windows%2C,with%20ROM%20bootloader%20to%20flash%20and%20erase%20fimware.

One thought crossing my mind is 'have you connected the SD card CS,MOSI,MISO,SCK,GND,VCC" correctly?
On 01/02/2023 14:04:56, EngineerTrooper ***@***.***> wrote:
reflash the UF2
I haven't done this before. Would be my first time. How I do that with my ESP32 Dev Kit C V2 ?
The Internet shows some tutorials but do you have a specific way in mind, @BNNorman [https://github.com/BNNorman] ?
—
Reply to this email directly, view it on GitHub [https://github.com/orgs/micropython/discussions/10581#discussioncomment-4840969], or unsubscribe [https://github.com/notifications/unsubscribe-auth/ADY5NXJJIJ2H5COZBAFHQ6DWVJUQRANCNFSM6AAAAAAUFPDZ5Q].
You are receiving this because you were mentioned.Message ID: ***@***.***>
[14fcd5d4-d920-46ac-9234-f760a27f16c0]

**Reply by engineertrooper** [reply]
> How did you put micropython on your esp32 before? Basically, use esptool. For doing it on Windows there's this tutorial  

I used esptool and also the commands in the tutorial. I reflashed fully the board yesterday to make sure it isn't corrupted. Also I tested v1.18, but error still comes up.

Also checked the pins twice or more times. If the cables aren't connected right I get error ```[0;31mE (120) sdmmc_sd: sdmmc_init_sd_if_cond: send_if_cond (1) returned 0x108[0m``` (I did that just to make sure everything is fine, and the hardware is connected correct.)

**Reply by robert-hh** [reply]
Using esptool.py to flash the firmware is fine.

**Comment by robert-hh** [comment]
So I took a plain vanilla ESP32 NodeMCU type, similar to what you use. Firmware version:
`MicroPython v1.19.1-849-g850f09b10 on 2023-02-01; ESP32 module with ESP32`
I wired up an external SD card module the same way as you did, inserted a 32G SanDisk SD card and ran the script:
```
import os
import machine

sd = machine.SDCard(slot=2)
vfs=os.VfsFat(sd)
os.mount(vfs, "/sd")  # mount

os.listdir('/sd')    # list directory contents
```
Everything works fine.

**Reply by engineertrooper** [reply]
I will order a new board now... Hope the error is going away then...

**Reply by robert-hh** [reply]
If you use an external SD, patch cables or bread boards may be a problem as well.

**Reply by engineertrooper** [reply]
Yes, it is external, but I used the hardware also with ArduinoIDE and ESP-IDF for both it works. If the hardware is not connected correctly I get the error ```[0;31mE (120) sdmmc_sd: sdmmc_init_sd_if_cond: send_if_cond (1) returned 0x108[0m```

I keep you updated

**Comment by BNNorman** [comment]
Has the SD card hdw worked on another device? OR perhaps use Arduino IDE to flash an SD card test program? If you have another Dev Kit that would be more convenient.

Sorry to ask such basic questions, I'm running out of ideas and clutching at straws.
On 01/02/2023 14:25:37, EngineerTrooper ***@***.***> wrote:
How did you put micropython on your esp32 before? Basically, use esptool. For doing it on Windows there's this tutorial
I used esptool and also the commands in the tutorial. I reflashed fully the board yesterday to make sure it isn't corrupted. Also I tested v1.18, but error still comes up.
Also checked the pins twice or more times. If the cables aren't connected right I get error [0;31mE (120) sdmmc_sd: sdmmc_init_sd_if_cond: send_if_cond (1) returned 0x108[0m (I did that just to make sure everything is fine, and the hardware is connected correct.)
—
Reply to this email directly, view it on GitHub [https://github.com/orgs/micropython/discussions/10581#discussioncomment-4841178], or unsubscribe [https://github.com/notifications/unsubscribe-auth/ADY5NXMN5CX7RB7QUM7C4YDWVJW6BANCNFSM6AAAAAAUFPDZ5Q].
You are receiving this because you were mentioned.Message ID: ***@***.***>
[89fa8b53-407f-46e1-916f-228719903429]

**Reply by engineertrooper** [reply]
I will order a new board, maybe a friend will borrow me his board today. Let's see. I keep you updated.

**Comment by engineertrooper** [comment, marked as answer]
My friend was giving me a TTGO-Koala Esp32 Board.
With ```MicroPython v1.18 on 2022-01-17; ESP32 module with ESP32```

tested the following code:

```python
import os
import machine

sd = machine.SDCard(slot=2)
vfs=os.VfsFat(sd)
os.mount(vfs, "/sd")  # mount

fn = open('/sd/textsd.txt', 'w')
fn.write('some data')
fn.close()

os.listdir('/sd')    # list directory contents
```

works like a charm!

-------------

Thank you for your help and ideas. I think my ESP32 Dev Kit C V2 has a broken arm or leg... don't know. Time for a new one.

I close this discussion as solved, now.

**Reply by cafalchio** [reply]
OSError: [Errno 19] ENODEV

**Reply by ADV1K** [reply]
I'm getting the same error, can't seem to figure it out.
@cafalchio did you figure it out?

**Comment by bingal** [comment]
I've encountered a similar issue before; eventually, switching the power supply from 3.3V to 5V did the trick even though the SD card module specifications state that it supports 3.3V.

**Reply by BNNorman** [reply]
Perhaps a large capacitor close to the SDcard VCC and GND might help.  I should have tried that before. Right now I can't test that idea.

**Comment by cafalchio** [comment]
@
Advik
Not yet, I assumed it was a hardware failure in my board.
Lets see if someone solves it

Matheus Cafalchio, PhD

**Reply by ADV1K** [reply]
Interesting, do you have any idea what might have caused it?
Also did you try to access the sdcard with arduino?

**Comment by ADV1K** [comment]
So I finally figured it out, MicroPython seems to only support **unpartitioned** sd cards, that is without any partition table, I also tried creating a partition table with only one primary partition but that didn't work.
If you're facing the same issue then you can format your card on linux like this (change sda with the block device of your card, find it using `lsblk`)

```shell
sudo mkfs.fat -F 32 /dev/sda
```

**Reply by lanxinwoaini** [reply]
Hello, can I simply execute sudo mkfs.fat -F 32 /dev/sdb? I've tried to do so, but I'm still encountering an error.

**Reply by projectgus** [reply]
You can also format the card from inside MicroPython if you need to:

```py
import vfs, machine
sd = machine.SDCard(...)
vfs.VfsFat.mkfs(sd)
```

**Reply by engineertrooper** [reply]
can't test that anymore. boards are gone. But thanks for these answers!

**Comment by cafalchio** [comment]
@advik Thank you very much for finding it.
I bet more people will come here to find it out.
Matheus Cafalchio, PhD

**Comment by nopthetop** [comment]
I am also having this issue. I know it's not hardware related because I was able to get this to work using circuit python.
